backend/src/main/java/com/bakdata/conquery/models/index/IndexService.java

```diff
@@ -0,0 +1,179 @@
+		this.csvParserSettings.setHeaderExtractionEnabled(true);
+	}
+
+	private final Queue<CompletableFuture<? extends Index<?,?>>> requestedMappings = new LinkedList<>();
+
+	private final LoadingCache<IndexKey<?,?>, Index<?,?>> mappings = CacheBuilder.newBuilder().build(new CacheLoader<IndexKey<?,?>, Index<?,?>>() {
+		@Override
+		public Index<?,?> load(@NotNull IndexKey key) throws Exception {
+			log.info("Started to parse mapping {}", key);
+
+			final Index<?,?> int2ext = key.createIndex();
+
+			final CsvParser csvParser = new CsvParser(csvParserSettings);
+
+			try (InputStream inputStream = key.getCsv().openStream()) {
+
+				final IterableResult<Record, ParsingContext> records = csvParser.iterateRecords(inputStream);
+
+				// Set default to "" for all columns
+				final String[] headers = records.getContext().headers();
+				records.getContext().recordMetaData().setDefaultValueOfColumns("", headers);
+
+				// Iterate records
+				for (Record row : records) {
+					final Pair<String, Map<String, String>> pair = computeInternalExternal(key, csvParser, row);
+					if (pair == null) {
+						continue;
+					}
+
+					final String internalValue = pair.getLeft();
+					final Map<String, String> externalValue = pair.getRight();
+
+					try {
+						int2ext.put(internalValue, externalValue);
+					} catch (IllegalArgumentException e) {
+						log.warn(
+								"Skipping mapping '{}'->'{}' in row {}, because there was already a mapping",
+								internalValue,
+								externalValue,
+								csvParser.getContext().currentLine(),
+								(Exception) (log.isTraceEnabled() ? e : null)
+						);
+					}
+				}
+			}
+			catch (IOException ioException) {
+				log.warn("Failed to open {}", key.getCsv(), ioException);
+				throw ioException;
+			}
+
+			log.info("Finished parsing mapping {} with {} entries", key, int2ext.size());
+
+			// Run finalizing operations on the index
+			int2ext.finalizer();
+
+			return int2ext;
+		}
+
+		@Nullable
+		private Pair<String, Map<String, String>> computeInternalExternal(@NotNull IndexKey<?, ?> key, CsvParser csvParser, Record row) {
+			final StringSubstitutor substitutor = new StringSubstitutor(row::getString, "{{", "}}", StringSubstitutor.DEFAULT_ESCAPE);
+
+			final String internalValue = row.getString(key.getInternalColumn());
+
+			if (internalValue == null) {
+				log.trace("Could not create a mapping for row {} because the cell for the internal value was empty. Row: {}", csvParser.getContext()
+																																	   .currentLine(),
+						  log.isTraceEnabled()
+						  ? StringUtils.join(row.toFieldMap())
+						  : null
+				);
+				return null;
+			}
+
+			final List<String> externalTemplates = key.getExternalTemplates();
+
+			/*
+			 Actually the size of the map is fixed.
+			 We allocate a bit more and set the
+			 load factor to avoid reallocation
+			 and rehashing
+			 */
+			final Map<String, String> templateToConcrete = new HashMap<>(externalTemplates.size() + 1, 1);
```

Comment:
Analog wie com.bakdata.conquery.util.CollectionsUtil#createSmallestSet allozieren?

Reply:
Ich denke ich kann direkt Object2ObjectArrayMap nehmen

Reply:
Die ist halt sehr langsam aber wenn das von der Performance reicht

Reply:
Ich habe gerade einen Benckmark in [µs] gemacht: Eine normale HashMap<>() ist insgesamt (schreiben + lesen) oft schneller gewesen, dann benutze ich sie auch:
| items         | ArrayMap<>(mapSize) | HashMap<>() | HashMap<>(mapSize +  1 ,  1) |
|---------------|--------------------:|------------:|-----------------------------:|
| insert 100    |                 414 |          25 |                          452 |
| insert 1_000  |               11113 |         200 |                          234 |
| insert 10_000 |              507691 |        2704 |                         6327 |
| insert 50_000 |             9923349 |        9768 |                        11051 |
| lookup 100    |                 368 |          25 |                            5 |
| lookup 1_000  |               19292 |         628 |                           56 |
| lookup 10_000 |              464687 |         542 |                         2437 |
| lookup 50_000 |            14573534 |        9761 |                        13200 |

Reply:
```
package com.bakdata.conquery;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.TimeUnit;
import java.util.stream.IntStream;

import com.google.common.base.Stopwatch;
import it.unimi.dsi.fastutil.objects.Object2ObjectArrayMap;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

@Slf4j
public class MapTest {

	private final Random random = new Random();

	public static int[] argumentsStream() {
		return new int[]{
				10,
				100,
				1_000,
				10_000,
				50_000};
	}

	@ParameterizedTest
	@MethodSource("argumentsStream")
	void testArrayMap(Integer mapSize) {
		doTestInsetAndLookup(mapSize, new Object2ObjectArrayMap<>(mapSize));
	}

	@ParameterizedTest
	@MethodSource("argumentsStream")
	void testHashMapOpen(Integer mapSize) {
		doTestInsetAndLookup(mapSize, new HashMap<>());
	}

	@ParameterizedTest
	@MethodSource("argumentsStream")
	void testHashMapClosed(Integer mapSize) {
		doTestInsetAndLookup(mapSize, new HashMap<>(mapSize + 1, 1));
	}

	private void doTestInsetAndLookup(int size, Map<String, String> map) {
		String[] strings = new String[size];
		IntStream.range(0, size).forEach(i -> {
			byte[] strBuffer = new byte[random.nextInt(100)];
			random.nextBytes(strBuffer);
			strings[i] = new String(strBuffer);
		});

		Stopwatch stopwatch = Stopwatch.createStarted();
		for (String string : strings) {
			map.put(string, string);
		}
		log.info("Done insert in {} µs", stopwatch.elapsed(TimeUnit.MICROSECONDS));

		String[] stringSorted = Arrays.stream(strings).sorted().toArray(String[]::new);

		int preventOptimization = 0;

		stopwatch = Stopwatch.createStarted();

		for (String lookup : stringSorted) {
			preventOptimization += map.get(lookup).length();
		}

		log.info("Done lookup in {} µs\t\t(optimization stopper: {})", stopwatch.elapsed(TimeUnit.MICROSECONDS), preventOptimization);
	}
}
```

Reply:
Dein Test ist vmtl nicht representativ weil du ja idR 3-4 Schlüssel haben solltest und die in den Cache passen sollten, aber ja das passt mit hashmap